### PR TITLE
[libc++][test] Fix simple warnings

### PR DESCRIPTION
**Pull request by StephanTLavavej**
Found while running libc++'s tests with MSVC's STL.

This contains 3 commits to fix simple warnings, and 1 cleanup commit.

* Add void-casts to fix `-Wunused-variable` warnings.
  + Something about MSVC's STL or our test environment makes Clang/LLVM emit warnings here. As the variables are intentionally unused, void-casting is the simplest way to acknowledge this.
* Avoid sign/truncation warnings in `ConvertibleToIntegral.h`.
  + As these warnings are emitted purely within test support code, we can silence them with `static_cast` without worrying about disrupting any usage. Examples:
    ```
    libcxx\test\std\containers\views\mdspan\layout_stride\../ConvertibleToIntegral.h(19): warning C4242: 'return': conversion from 'const int' to 'unsigned char', possible loss of data        
    libcxx\test\std\containers\views\mdspan\layout_stride\../ConvertibleToIntegral.h(20): warning C4242: 'return': conversion from 'const int' to 'signed char', possible loss of data
    libcxx\test\std\containers\views\mdspan\layout_stride\../ConvertibleToIntegral.h(32): warning C4242: 'return': conversion from 'int' to 'char', possible loss of data
    ```
* Add `TEST_STD_AT_LEAST_23_OR_RUNTIME_EVALUATED` to avoid mixing preprocessor and runtime tests.
  + When MSVC sees `if (TEST_STD_VER >= 23 || !TEST_IS_CONSTANT_EVALUATED)` it emits `warning C4127: conditional expression is constant` because the first part is truly a compile-time constant. To fix this, I'm mechanically simplifying it into a centralized macro.
* Cleanup: Add `TEST_STD_AT_LEAST_20_OR_RUNTIME_EVALUATED` for consistency.
  + The codebase had the same pattern, but with C++20. Although these didn't seem to be causing warnings for MSVC's STL, using the new pattern makes it more likely that it'll be followed for other code (including C++26 and beyond).

### CODE REVIEW

File: libcxx/test/std/algorithms/alg.modifying.operations/alg.move/move.pass.cpp [hunk 1]
```diff
@@ -94,8 +94,13 @@ struct Test1OutIters {
 
 TEST_CONSTEXPR_CXX20 bool test() {
   types::for_each(types::cpp17_input_iterator_list<int*>(), TestOutIters());
-  if (TEST_STD_VER >= 23 || !TEST_IS_CONSTANT_EVALUATED)
+
+#if TEST_STD_VER < 23
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply libcxx/test/std/algorithms/alg.modifying.operations/alg.move/move.pass.cpp]: Done! I figured out a clean way to do this without introducing a new helper function.

**Comment by StephanTLavavej** [comment]
@ldionne I've introduced a new macro for testing `constexpr`-in-C++23 code that avoids the need for complicated conditionals.

**Comment by StephanTLavavej** [comment]
Ok, I've got an even simpler approach that should work.

The history of reverts and reworks was getting too complicated, so I tidied it up, rebased, and force-pushed. Apologies if this complicates incremental reviewing (I think it should hopefully be easier to follow). Also updated the PR description.

**Review by ldionne — APPROVED** [review]
Not a huge fan of the new macros because they won't compose super well, but it'll do. Thanks for the fix!

**Pull request merged**